src/index.js

```diff
@@ -24,27 +24,76 @@ const WATCHING = 'WATCHING'
 const NOT_IGNORED = 'NOT_IGNORED'
 const KICKED = 'KICKED'
 const PAUSED = 'PAUSED'
+const SOCKET_EVENTS = ['connect', 'close', 'error', 'ready'];
 
 module.exports = JackdClient
 
 function JackdClient() {
   this.socket = new Socket()
   this.socket.setEncoding('ascii')
+  this.connected = false
+
+  this.socket.on('ready', () => {
+    this.connected = true
+  })
+
+  this.socket.on('close', () => {
+    this.connected = false
+  })
+}
+
+/**
+ * Useful check for environments where network partitioning is common.
+ * @returns {boolean}
+ */
+JackdClient.prototype.isConnected = function() {
+  return this.connected;
+}
+
+/**
+ * Support event listeners on the underlying socket to support external
+ * reconnection logic.
+ * @param event
+ * @param listener
+ */
+JackdClient.prototype.on = function(event, listener) {
+  if (SOCKET_EVENTS.includes(event)) {
+    this.socket.on(event, listener);
+  }
+}
+
+/**
+ * Support one-time event listeners on the underlying socket to support
+ * external reconnection logic.
+ * @param event
+ * @param listener
+ */
+JackdClient.prototype.once = function(event, listener) {
+    if (SOCKET_EVENTS.includes(event)) {
+        this.socket.once(event, listener);
+    }
 }
 
 JackdClient.prototype.connect = async function() {
   const socket = this.socket
-  let host, port
+  let host = undefined, port = 11300
 
   if (arguments.length === 1) {
     const [opts] = arguments
     host = opts.host
     port = opts.port
   }
 
-  await new Promise(resolve =>
-    socket.connect(port || 11300, host || undefined, resolve)
-  )
+  await new Promise((resolve, reject) => {
+      socket.once('error', (error) => {
+        if (error.code === 'EISCONN') {
+          return resolve();
```

Comment:
What's going on here? An error that resolves the promise?

Reply:
This error indicates the socket is already connected. I actually hit this a couple of times in testing, and the reconnect logic would reject endlessly once it got into this state. A different approach would be to let the catch block of reconnect evaluate the type of error, but since this operation is "connect", it seemed to me that the client is in the desired state so it shouldn't be an error.

Reply:
Interesting - I wouldn't have guessed that. Thanks.